urld58fd5e7a9_longrm_weibo_robot_tgz-pJ8-weibo4j_org_json_TestJ8/src/weibo4j/http/HttpClient.java

```diff
@@ -0,0 +1,963 @@
+        return httpRequest(url, null, false);
+    }
+
+    protected Response httpRequest(String url, PostParameter[] postParams,
+            boolean authenticated) throws WeiboException {
+    	//统一增加source 参数
+		int len = 1;
+		PostParameter[] newPostParameters = postParams;
+    	String method = "GET";
+    	if (postParams != null) {
+    		method = "POST";
+			len = postParams.length + 1;
+			newPostParameters = Arrays.copyOf(postParams, len);
+			newPostParameters[postParams.length] = new PostParameter("source",
+					Weibo.CONSUMER_KEY);
+    	}
+    	return httpRequest(url, newPostParameters, authenticated, method);
+    }
+
+    public Response httpRequest(String url, PostParameter[] postParams,
+                                 boolean authenticated, String httpMethod) throws WeiboException {
+        int retriedCount;
+        int retry = retryCount + 1;
+        Response res = null;
+        for (retriedCount = 0; retriedCount < retry; retriedCount++) {
+            int responseCode = -1;
+            try {
+                HttpURLConnection con = null;
+                OutputStream osw = null;
+                try {
+                    con = getConnection(url);
+                    con.setDoInput(true);
+                    setHeaders(url, postParams, con, authenticated, httpMethod);
+                    if (null != postParams || "POST".equals(httpMethod)) {
+                        con.setRequestMethod("POST");
+                        con.setRequestProperty("Content-Type",
+                                "application/x-www-form-urlencoded");
+                        con.setDoOutput(true);
+                        String postParam = "";
+                        if (postParams != null) {
+                        	postParam = encodeParameters(postParams);
+                        }
+                        log("Post Params: ", postParam);
+                        byte[] bytes = postParam.getBytes("UTF-8");
+
+                        con.setRequestProperty("Content-Length",
+                                Integer.toString(bytes.length));
+                        
+                        osw = con.getOutputStream();
+                        osw.write(bytes);
+                        osw.flush();
+                        osw.close();
+                    } else if ("DELETE".equals(httpMethod)){
+                        con.setRequestMethod("DELETE");
+                    } else {
+                        con.setRequestMethod("GET");
+                    }
+                   
+                   
+                    res = new Response(con);
+                    responseCode = con.getResponseCode();
+                    if(DEBUG){
+                        log("Response: ");
+                        Map<String, List<String>> responseHeaders = con.getHeaderFields();
+                        for (String key : responseHeaders.keySet()) {
+                            List<String> values = responseHeaders.get(key);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.